### PR TITLE
add new converter module

### PR DESCRIPTION
**Pull request by johnnv1**
Related to #30

### CODE REVIEW

File: lapixdl/formats/annotation.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+
+@dataclass
+class Annotation:
+    geometry: Polygon | MultiPolygon
+    category_id: int
+    iscrowd: int = 0
+
+    @property
+    def bbox(self) -> BBox:
```
Comment (Andrevmatias) [inline lapixdl/formats/annotation.py:161]:
```suggestion
    def to_bbox(self) -> BBox:
```

File: lapixdl/formats/annotation.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        """
+
+        return self.area + bbox.area - (intersection_area or self.intersection_area_with(bbox))
+
+    def to_polygon(self) -> Polygon:
+        return Polygon(self.coords)
+
+
+@dataclass
+class Annotation:
```
Comment (Andrevmatias) [inline lapixdl/formats/annotation.py:155]:
```suggestion
class Segmentation:
```
or
```suggestion
class SegmentationInstance:
```

File: lapixdl/formats/annotation.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    @property
+    def bbox(self) -> BBox:
+        return bounds_to_BBox(self.geometry.bounds, self.category_id)
+
+    @property
+    def _geo_type(self) -> str:
+        return self.geometry.geom_type
+
+    @property
+    def coco_bbox(self) -> list[float | int]:
```
Comment (Andrevmatias) [inline lapixdl/formats/annotation.py:169]:
```suggestion
    def to_coco_bbox(self) -> list[float | int]:
```

File: lapixdl/formats/lapix.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+from typing import Any
+
+import pandas as pd
+import shapely.wkt
+
+from lapixdl.base import FileTypeError
+
+
+class Lapix(pd.DataFrame):
```
Comment (Andrevmatias) [inline lapixdl/formats/lapix.py:11]:
```suggestion
class LapixDataFrame(pd.DataFrame):
```

File: lapixdl/formats/lapix.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    _metadata = ['geometry', 'category']
+
+    @property
+    def _constructor(self):
+        return Lapix
+
+
+def load(
```
Comment (Andrevmatias) [inline lapixdl/formats/lapix.py:19]: Why are the `load`(static) and `save` methods outside the Lapix class?

File: lapixdl/formats/labelbox.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    raw: list[dict[str, Any]]
+) -> bool:
+    if not isinstance(raw, list):
+        raise TypeError('A list of dictionaries representing the raw data from LabelBox was expected.')
+
+    expected_data = set({'ID', 'External ID', 'Reviews', 'Label'})
+    for it in raw:
+        if not all(i in it for i in expected_data):
+            if 'Skipped' not in it:
+                raise KeyError(f'Not found expected values need to have `Skipped` or {expected_data}')
```
Comment (Andrevmatias) [inline lapixdl/formats/labelbox.py:30]:
```suggestion
                raise KeyError(f'The annotation values need to have `Skipped` or all of {expected_data} keys.')
```

File: lapixdl/formats/labelbox.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    return pd.DataFrame(raw)
+
+
+def validate(
+    raw: list[dict[str, Any]]
+) -> bool:
+    if not isinstance(raw, list):
+        raise TypeError('A list of dictionaries representing the raw data from LabelBox was expected.')
+
+    expected_data = set({'ID', 'External ID', 'Reviews', 'Label'})
```
Comment (Andrevmatias) [inline lapixdl/formats/labelbox.py:26]: This is a constant value, declare it out of the method.

File: lapixdl/formats/labelbox.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    for it in raw:
+        if not all(i in it for i in expected_data):
```
Comment (Andrevmatias) [inline lapixdl/formats/labelbox.py:28]:
```suggestion
        if 'Skipped' not in it and any(i not in it for i in expected_data):
```

File: lapixdl/convert.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+    if geo.geom_type == 'Polygon':
+        return [coco_pol(geo)]
+    elif geo.geom_type == 'MultiPolygon':
+        return [coco_pol(g) for g in geo.geoms]
+    else:
+        raise TypeError(f'Geometry shape is not a polygon or MultiPolygon. This is a {geo.geom_type}.')
+
+
+def __to_OD_COCO_annotations(
```
Comment (Andrevmatias) [inline lapixdl/convert.py:169]:
```suggestion
def __to_od_coco_annotations(
```
the pattern for method names in this project seems to be `lower_case`

File: lapixdl/convert.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    annotations_coco = to_OD_COCO_annotations(lapix_df, decimals, processes)
+
+    coco_od = {
+        'categories': categories_coco,
+        'images': images_coco,
+        'annotations': annotations_coco,
+        'info': info_coco if isinstance(info_coco, dict) else {}
+    }
+
+    return coco_od
```
Comment (Andrevmatias) [inline lapixdl/convert.py:251]: I think it's better to have `x_to_y` functions to improve usability (i.e. `labelbox_to_coco`). Even if it's only a shortcut to `(x) -> lapix_to_y(x_to_lapix(x))`.

File: lapixdl/base.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+from __future__ import annotations
+
+import os
+
+
+class FileTypeError(RuntimeError):
+    pass
+
+
+def basename(
```
Comment (Andrevmatias) [inline lapixdl/base.py:10]: Isn't it better to just use the Pathlib

File: README.md [hunk 12]
```diff
@@ -87,27 +87,39 @@ from lapixdl.evaluation.visualize import show_detections
 The available color maps are the [ones from matplotlib](https://matplotlib.org/3.1.1/gallery/color/colormap_reference.html).
 
 ### For Data Conversion
-
-This module exports the following functions for data conversion:
-```python
-from lapixdl.convert.labelbox import labelbox_to_coco
-```
+This module exports the functions for data conversio.
```
Comment (Andrevmatias) [inline README.md:90]: Why did you remove the list of exported functions?

File: tests/convert/coco_expect.json [hunk 13]
```diff
@@ -1,1316 +0,0 @@
-{
```
Comment (Andrevmatias) [inline tests/convert/coco_expect.json:1]: Isn't it better to have "real-world examples" for at least integration tests?

**Review by Andrevmatias — CHANGES_REQUESTED** [review]
Just some points to think about. Also, until now, all the exported methods and classes have documentation. I think it's better to keep this pattern.

File: lapixdl/formats/lapix.py (hunk 5 above)
Comment (johnnv1) [reply lapixdl/formats/lapix.py:19]: I think the `LapixDataFrame` will be used just inside the library, so nobody will explicitly use it.

 Thinking on this, for me, the code will be cleaner like this:
```python
from lapixdl.formats import lapix

lapix.load(...)
```

That this way:
```python
from lapixdl.formats.lapix import LapixDataFrame

LapixDataFrame.load(...)

# or 

from lapixdl.formats import lapix


lapix.LapixDataFrame.load(...)
```

Also, at ccagt-utils I start using as methods, and I ended up thinking that this brings an extra level of paradigm without any benefit

File: lapixdl/base.py (hunk 11 above)
Comment (johnnv1) [reply lapixdl/base.py:10]: I really don't like the Pathlib pattern for join (`/`), using the `os` functions seems clearer to me

File: README.md (hunk 12 above)
Comment (johnnv1) [reply README.md:90]: Because I change the way of the converter works, and don't know if I like the part of have the function `create_COCO_OD`... and I forgot to add it to the readme

File: tests/convert/coco_expect.json (hunk 13 above)
Comment (johnnv1) [reply tests/convert/coco_expect.json:1]: I think we can break it into small tests, seems to become clearer to me. By now, I just complete the coverage, but need improvements, to make sure that everything is being tested correctly.

File: lapixdl/formats/annotation.py (hunk 1 above)
Comment (johnnv1) [reply lapixdl/formats/annotation.py:161]: I add this class, but, isn't used in anywhere now, so I will just remove it. 

It is added with an idea of interface when unpacking annotations from the data frame... when this be usefully I add this back

File: lapixdl/formats/lapix.py (hunk 5 above)
Comment (Andrevmatias) [reply lapixdl/formats/lapix.py:19]: when you read it alone, `lapix.load(...)` does not tell you what is being loaded, while `LapixDataFrame.load(...)` does. Besides, you can import the `load` function alone and use `load(...)`, which makes it even more "obscure" what is happening...

File: tests/convert/coco_expect.json (hunk 13 above)
Comment (Andrevmatias) [reply tests/convert/coco_expect.json:1]: Until then, I think it's better to keep the old tests...

File: lapixdl/formats/annotation.py (hunk 2 above)
Comment (johnnv1) [reply lapixdl/formats/annotation.py]: Class removed on [3e30be4](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/3e30be492800341fe79636e124089e2425f45d21)

File: lapixdl/formats/annotation.py (hunk 3 above)
Comment (johnnv1) [reply lapixdl/formats/annotation.py]: Class removed on [3e30be4](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/3e30be492800341fe79636e124089e2425f45d21)

File: lapixdl/formats/lapix.py (hunk 4 above)
Comment (johnnv1) [reply lapixdl/formats/lapix.py]: Done on [3e30be4](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/3e30be492800341fe79636e124089e2425f45d21)

File: lapixdl/convert.py (hunk 9 above)
Comment (johnnv1) [reply lapixdl/convert.py]: Done on [3e30be4](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/3e30be492800341fe79636e124089e2425f45d21)

File: lapixdl/convert.py (hunk 10 above)
Comment (johnnv1) [reply lapixdl/convert.py:251]: Done on [3e30be4](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/3e30be492800341fe79636e124089e2425f45d21)

File: tests/convert/coco_expect.json (hunk 13 above)
Comment (johnnv1) [reply tests/convert/coco_expect.json:1]: Add an extra test with the real data from old tests on [48ef581](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/48ef581e6a26e7d373a2752c596b7b9fb339ca09)

File: lapixdl/formats/labelbox.py (hunk 6 above)
Comment (johnnv1) [reply lapixdl/formats/labelbox.py]: Done at [44c0d3d](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/44c0d3d2d2989da67155da2cfac9f84885ea453e)

File: lapixdl/formats/labelbox.py (hunk 7 above)
Comment (johnnv1) [reply lapixdl/formats/labelbox.py]: Done at [44c0d3d](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/44c0d3d2d2989da67155da2cfac9f84885ea453e)

File: lapixdl/formats/labelbox.py (hunk 8 above)
Comment (johnnv1) [reply lapixdl/formats/labelbox.py]: Done at [44c0d3d](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/44c0d3d2d2989da67155da2cfac9f84885ea453e)

File: lapixdl/convert.py (hunk 10 above)
Comment (johnnv1) [reply lapixdl/convert.py:300]: Add labelbox_to_coco on [ec63a4d](https://github.com/lapix-ufsc/lapixdl/pull/34/commits/ec63a4d36a1d016cff15ce7a92b123527a448f12)

**Review by Andrevmatias — APPROVED** [review]
